### PR TITLE
Use linux-fx-version solely and require linux runtime

### PR DESCRIPTION
**Pull request by LukaszStem**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 1]
```diff
@@ -309,14 +323,15 @@ def delete_connection_strings(resource_group_name, name, setting_names, slot=Non
...
-                              'DOCKER_REGISTRY_SERVER_PASSWORD', 'DOCKER_CUSTOM_IMAGE_NAME']
+                              'DOCKER_REGISTRY_SERVER_PASSWORD']
 APPSETTINGS_TO_MASK = ['DOCKER_REGISTRY_SERVER_PASSWORD']
 
 
 def update_container_settings(resource_group_name, name, docker_registry_server_url=None,
                               docker_custom_image_name=None, docker_registry_server_user=None,
                               docker_registry_server_password=None, slot=None):
     settings = []
+    returnVal = ""
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Please rename to `return_val`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 2]
```diff
@@ -181,7 +185,17 @@ def _fill_ftp_publishing_url(webapp, resource_group_name, name, slot=None):
...
+
+
+def _delete_linux_fx_version(resource_group_name, name):
+    fx_version = ''
+    return update_site_configs(resource_group_name, name, linux_fx_version=fx_version)
+
+
+def _get_linux_fx_version(resource_group_name, name):
+    site_config = get_site_configs(resource_group_name, name)
+    return str(site_config.linux_fx_version)
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: What is the type of `site_config.linux_fx_version` which requires a conversion to `str`. I assume it is a string type already, no?

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -334,13 +349,15 @@ def update_container_settings(resource_group_name, name, docker_registry_server_
...
     if docker_registry_server_password is not None:
         settings.append('DOCKER_REGISTRY_SERVER_PASSWORD=' + docker_registry_server_password)
-
     if docker_custom_image_name is not None:
-        settings.append('DOCKER_CUSTOM_IMAGE_NAME=' + docker_custom_image_name)
-        _add_linux_fx_version(resource_group_name, name, docker_custom_image_name)
-    update_app_settings(resource_group_name, name, settings, slot)
-    settings = get_app_settings(resource_group_name, name, slot)
-    return _mask_creds_related_appsettings(_filter_for_container_settings(settings))
+        returnVal += str(_add_linux_fx_version(resource_group_name, name, docker_custom_image_name))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: I don't think we need to convert it to `str`, cli's core output has pretty good logics to display everything

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 4]
```diff
@@ -334,13 +349,15 @@ def update_container_settings(resource_group_name, name, docker_registry_server_
...
+
+    if docker_registry_server_user or docker_registry_server_password or docker_registry_server_url:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: It is fine to condition on provided values.  I suggest we put a condition at the method's entry to error out if no settings are provided, then we can remove the condition here.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 5]
```diff
@@ -334,13 +349,15 @@ def update_container_settings(resource_group_name, name, docker_registry_server_
...
+        update_app_settings(resource_group_name, name, settings, slot)
+        settings = get_app_settings(resource_group_name, name, slot)
+        returnVal += str(_mask_creds_related_appsettings(_filter_for_container_settings(settings)))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Again, we don't need a conversion here.
Also you can add a new field with the value of the `docker_custom_image_name ` to get the information complete

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 6]
```diff
@@ -365,12 +382,14 @@ def _get_acr_cred(registry_name):
...
 
 def delete_container_settings(resource_group_name, name, slot=None):
+    _delete_linux_fx_version(resource_group_name, name)
     delete_app_settings(resource_group_name, name, CONTAINER_APPSETTING_NAMES, slot)
 
 
 def show_container_settings(resource_group_name, name, slot=None):
     settings = get_app_settings(resource_group_name, name, slot)
-    return _mask_creds_related_appsettings(_filter_for_container_settings(settings))
+    return "'docker-custom-image-name': {} {}".format(_get_linux_fx_version(resource_group_name, name),
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: You should update the dictionary, rather do string format which will break the output on non json mode

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 1 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: should we output a deprecating warning that "--docker-custom-image-name" will be renamed to a more appropriate name?

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 7]
```diff
@@ -179,9 +183,28 @@ def _fill_ftp_publishing_url(webapp, resource_group_name, name, slot=None):
     return webapp
 
 
+def _format_linux_fx_version(custom_image_name):
+    # set it as a space (Once fixed in backend, will actually set to empty here)
+    fx_version = ' '
+    # handles case of only spaces
+    if custom_image_name.strip():
+        fx_version = '{}|{}'.format('DOCKER', custom_image_name)
```
Comment (LukaszStem) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Yes. In the event of someone using ACR this logic is correct.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 8]
```diff
@@ -179,9 +183,29 @@ def _fill_ftp_publishing_url(webapp, resource_group_name, name, slot=None):
     return webapp
 
 
+def _format_linux_fx_version(custom_image_name):
+    # set it as a space (Once fixed in backend, will actually set to empty here)
+    fx_version = ' '
+    # handles case of only spaces
+    if custom_image_name.strip():
+        if 'DOCKER|' not in custom_image_name:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: could you please sync up with @ahmedelnably  to see whether this is right? You know the image name from ACR would be `myacr.azurecr.io/image-name:latest` and this rendering will make it very much wrong. If we don't have reliable mechanism, maybe we should just leave it untouched

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 9]
```diff
@@ -179,9 +183,29 @@ def _fill_ftp_publishing_url(webapp, resource_group_name, name, slot=None):
...
+
 def _add_linux_fx_version(resource_group_name, name, custom_image_name):
-    fx_version = '{}|{}'.format('DOCKER', custom_image_name)
-    update_site_configs(resource_group_name, name, linux_fx_version=fx_version)
+    fx_version = _format_linux_fx_version(custom_image_name)
+    return update_site_configs(resource_group_name, name, linux_fx_version=fx_version)
+
+
+def _delete_linux_fx_version(resource_group_name, name):
+    fx_version = _format_linux_fx_version('')
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: i don't think we need to format to clean up a value

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 8 above)
Comment (ahmedelnably) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: @yugangw-msft Yes, this is how it is in the backend, all custom Docker images are represented in Linux_FX_version as "DOCKER|<image name>"

Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Okay, so for ACR based image, what would it looks like?

Comment (ahmedelnably) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: DOCKER|<registryname>.azureacr.io/<imagename>:<tag>

Comment (LukaszStem) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: E.g. `DOCKER|myacr.azurecr.io/image-name:latest`

Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Is it case sensitive?

Comment (ahmedelnably) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: @LukaszStem to confirm.

Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: @LukaszStem, please also update the code using `.startswith`, `in` is not accurate

Comment (LukaszStem) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Lowercase `docker|` works. Will change to .startswith

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 9 above)
Comment (LukaszStem) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Fixed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 10]
```diff
@@ -179,9 +183,29 @@ def _fill_ftp_publishing_url(webapp, resource_group_name, name, slot=None):
     return webapp
 
 
+def _format_linux_fx_version(custom_image_name):
+    fx_version = custom_image_name
+    fx_version_lower = fx_version.lower()
+    # handles case of only spaces
+    if custom_image_name.strip():
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Any reason we need to handle white spaces? Values from command arguments should have them stripped by the command parser. 
If we do need to handle, then you need to use stripped version to construct the `fx_version` below.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 11]
```diff
@@ -54,12 +54,16 @@ def create_webapp(resource_group_name, name, plan, runtime=None, startup_file=No
     if is_linux:
         if runtime and deployment_container_image_name:
             raise CLIError('usage error: --runtime | --deployment-container-image-name')
-        if startup_file or runtime:
+        if startup_file:
             site_config.app_command_line = startup_file
+
+        if runtime:
             site_config.linux_fx_version = runtime
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: should we also apply `_format_linux_fx_version` here?

Comment (LukaszStem) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: We assume this runtime is a one shown from running `list-runtimes`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 10 above)
Comment (LukaszStem) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: If the custom-image name is a number of spaces, we want fx_version to become `' '`

**Comment by robertlagrant** [comment]
@yugangw-msft appreciate that this is 2.5 years old, but is there any documentation on what should go in the `linux-fx-version` field? I've only found pretty sporadic stuff. My particular use car is a Consumption plan Python function.

**Comment by yugangw-msft** [comment]
I have left Azure CLI team. Add @panchagnula to help

**Pull request merged**